lib/allocator/allocator.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package allocator manages allocation of resources.
+// Modified from https://github.com/kubernetes/kubernetes/blob/v1.15.12/pkg/registry/core/service/allocator/allocator.go
+package allocator
```

Comment:
I wonder if we should place it in lib/utils rather than in just lib.

Reply:
`utils` is a collection of disparate implementations with no single purpose. Usually, packages with generic names like this are discouraged in favor of more specific ones. `Allocator` in this context is a generic resource allocator with a specific purpose which I believe has its place at the top of the packages directory.